### PR TITLE
once: add package 'once'

### PR DESCRIPTION
**Pull request by hdonnay**
Provides for a construct like a sync.Once, but caring about whether
the function succeeded.

### CODE REVIEW

File: once/once.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	o.L.Lock()
+	defer o.L.Unlock()
+
+Wait:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```
Comment (kodabb) [inline once/once.go]: please document the return value

File: once/once.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	atomic.StoreUint32(&o.done, 1)
+	o.Broadcast()
+	return nil
+}
+
+// Invoke does bad pointer-to-interface things to always return an error.
+func (o *Once) invoke(f func() error) (err error) {
+	defer func(e *error) {
```
Comment (kodabb) [inline once/once.go]: add a note about this magic

File: once/once.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+// Success is an object that will perform exactly one action if successful.
+type Success struct {
+	// This is an atomic instead of, say, a bool so that callers can hot-path without acquiring a lock.
+	done uint32
+
+	// This Cond protects the running bool, so that only one Do() execution is happening at once.
+	sync.Cond
+	running bool
```
Comment (kodabb) [inline once/once.go]: I'd add one line of documentation even if it's evident by the name

File: once/once.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	defer o.L.Unlock()
+
+	for o.running {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		o.Wait()
+	}
+	o.running = true
+	defer func() { o.running = false }()
```
Comment (kodabb) [inline once/once.go]: nit: it's fine to unroll this on multiple lines

File: once/once.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	return o
+}
+
+// Do calls the function f if and only if Do is being called for the
+// first time for this instance of Once and previous calls were not successful.
+// In other words, given
+//     var once *Once = New()
+// if once.Do(ctx, f) is called multiple times, f will be invoked until it returns
+// a non-nil error, even if f has a different value in each invocation.
+// A new instance of Once is required for each function to execute.
```
Comment (kodabb) [inline once/once.go]: s/Once/Success/

File: once/once.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+// Do is intended for initialization that must be run exactly once if successful.
+//
+// Because no call to Do returns until the one call to f returns, if f causes
+// Do to be called, it will deadlock.
+//
+// If f panics, Do considers it to have returned with an error, so future calls
+// of Do will invoke f again.
+//
+// If the context has been canceled before f is called successfully,
+// context.Canceled will be returned
```
Comment (kodabb) [inline once/once.go]: "... will be returned. Callers are responsible to gracefully handle this event."

File: once/once.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	defer func() { o.running = false }()
+
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+	if atomic.LoadUint32(&o.done) != 0 {
+		return nil
+	}
+
+	if err := o.invoke(f); err != nil {
```
Comment (kodabb) [inline once/once.go]: maybe mention that we call Signal to wake up a single routine

File: once/once.go (hunk 3 above)
Comment (hdonnay) [inline once/once.go]: ok

File: once/once.go (hunk 4 above)
Comment (hdonnay) [inline once/once.go]: ok

File: once/once.go (hunk 5 above)
Comment (hdonnay) [inline once/once.go]: ok

File: once/once.go (hunk 6 above)
Comment (hdonnay) [inline once/once.go]: ok

File: once/once.go (hunk 7 above)
Comment (hdonnay) [inline once/once.go]: ok

File: once/once.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// Success is an object that will perform exactly one action if successful.
+type Success struct {
+	// This is an atomic instead of, say, a bool so that callers can hot-path without acquiring a lock.
+	done uint32
+
```
Comment (kodabb) [inline once/once.go]: nit: drop the extra line

**Pull request merged**